drivers/ieee802154/ieee802154_dw1000.c

```diff
@@ -0,0 +1,1680 @@
+	u64_t tmp_fs;
+	u32_t tx_fctrl;
+	u8_t sys_ctrl = DWT_SYS_CTRL_TXSTRT;
+
+	if (atomic_test_and_set_bit(&ctx->state, DWT_STATE_TX)) {
+		LOG_ERR("Transceiver busy");
+		return -EBUSY;
+	}
+
+	k_sem_reset(&ctx->phy_sem);
+	k_sem_take(&ctx->spi_sem, K_FOREVER);
+
+	switch (tx_mode) {
+	case IEEE802154_TX_MODE_DIRECT:
+		break;
+	case IEEE802154_TX_MODE_TXTIME:
+		/*
+		 * tx_time is the high 32-bit of the 40-bit system
+		 * time value at which to send the message.
+		 */
+		txts = net_pkt_timestamp(pkt);
+		tmp_fs = txts->second * NSEC_PER_SEC + txts->nanosecond;
+		tmp_fs *= 1000U * 1000U;
+
+		tx_time = (tmp_fs / DWT_TS_TIME_UNITS_FS) >> 8;
+		sys_ctrl |= DWT_SYS_CTRL_TXDLYS;
+		/* DX_TIME is 40-bit register */
+		dwt_reg_write_u32(ctx, DWT_DX_TIME_ID, 1, tx_time);
+
+		LOG_DBG("ntx hi32 %x", tx_time);
+		LOG_DBG("sys hi32 %x",
+			dwt_reg_read_u32(ctx, DWT_SYS_TIME_ID, 1));
+		break;
+	default:
+		LOG_ERR("TX mode %d not supported", tx_mode);
+		goto error;
+	}
+
+	LOG_HEXDUMP_DBG(frag->data, len, "TX buffer:");
+
+	/*
+	 * See "3 Message Transmission" in DW1000 User Manual for
+	 * more details about transmission configuration.
+	 */
+	if (dwt_register_write(ctx, DWT_TX_BUFFER_ID, 0, len, frag->data)) {
+		LOG_ERR("Failed to write TX data");
+		goto error;
+	}
+
+	tx_fctrl = dwt_reg_read_u32(ctx, DWT_TX_FCTRL_ID, 0);
+	/* Clear TX buffer index offset, frame length, and length extension */
+	tx_fctrl &= ~(DWT_TX_FCTRL_TFLEN_MASK | DWT_TX_FCTRL_TFLE_MASK |
+		      DWT_TX_FCTRL_TXBOFFS_MASK);
+	/* Set frame length and ranging flag */
+	tx_fctrl |= (len + DWT_FCS_LENGTH) & DWT_TX_FCTRL_TFLEN_MASK;
+	tx_fctrl |= DWT_TX_FCTRL_TR;
+	/* Update Transmit Frame Control register */
+	dwt_reg_write_u32(ctx, DWT_TX_FCTRL_ID, 0, tx_fctrl);
+
+	dwt_disable_txrx(ctx);
+
+	/* Begin transmission */
+	dwt_reg_write_u8(ctx, DWT_SYS_CTRL_ID, DWT_SYS_CTRL_OFFSET, sys_ctrl);
+
+	if (sys_ctrl & DWT_SYS_CTRL_TXDLYS) {
+		u32_t sys_stat = dwt_reg_read_u32(ctx, DWT_SYS_STATUS_ID, 0);
+
+		if (sys_stat & DWT_SYS_STATUS_HPDWARN) {
+			LOG_WRN("Half Period Delay Warning");
+		}
+	}
+
+	k_sem_give(&ctx->spi_sem);
+	/* Wait for the TX confirmed event */
+	k_sem_take(&ctx->phy_sem, K_FOREVER);
+
+	if (IS_ENABLED(CONFIG_NET_PKT_TIMESTAMP)) {
+		u8_t ts_buf[sizeof(u64_t)] = {0};
+		struct net_ptp_time timestamp;
+
+		k_sem_take(&ctx->spi_sem, K_FOREVER);
+		dwt_register_read(ctx, DWT_TX_TIME_ID,
+				  DWT_TX_TIME_TX_STAMP_OFFSET,
+				  DWT_TX_TIME_TX_STAMP_LEN,
+				  ts_buf);
+		LOG_DBG("ts  hi32 %x", (u32_t)(sys_get_le64(ts_buf) >> 8));
+		LOG_DBG("sys hi32 %x",
+			dwt_reg_read_u32(ctx, DWT_SYS_TIME_ID, 1));
+		k_sem_give(&ctx->spi_sem);
+
+		tmp_fs = sys_get_le64(ts_buf) * DWT_TS_TIME_UNITS_FS;
+		timestamp.second = (tmp_fs / 1000000) / NSEC_PER_SEC;
+		timestamp.nanosecond = (tmp_fs / 1000000) % NSEC_PER_SEC;
+		net_pkt_set_timestamp(pkt, &timestamp);
+	}
+
+	atomic_clear_bit(&ctx->state, DWT_STATE_TX);
+
+	if (atomic_test_bit(&ctx->state, DWT_STATE_RX_DEF_ON)) {
```

Comment:
Shouldn'd this be done as well on tx error case? Because currently if tx failed, you will not re-enable rx though it should be on by default. (that's my understanding at least).

Reply:
It would not call `dwt_disable_txrx(ctx)`; and thus not disable reception.

Reply:
indeed line 852